src/timesteppers.jl

```diff
@@ -49,4 +49,90 @@ function update_time(s::FerriteSolver{<:Any,<:FixedTimeStepper}, t, step, conver
         throw(ConvergenceError(msg))
     end
     return s.timestepper.t[step+1], step+1
-end
\ No newline at end of file
+end
+
+
+"""
+    AdaptiveTimeStepper(
+        Δt_init::T, t_end::T; 
+        t_start=zero(T), Δt_min=Δt_init, Δt_max=typemax(T), 
+        change_factor=T(1.5), num_converged_to_increase::Int=1) where T
+
+An adaptive time stepper, starting with a step `Δt_init` and with a 
+maximum time of `t_end`. The `change_factor` describes how much the 
+time step is changed in the case of (a) no convergence: 
+`Δt/=change_factor` and (b) converged the last `num_converged_to_increase`
+time steps: `Δt*=change_factor`.
+"""
+struct AdaptiveTimeStepper{T}
+    t_start::T
+    t_end::T
+    Δt_init::T
+    Δt_min::T
+    Δt_max::T
+    change_factor::T 
+    num_converged_to_increase::Int
+    num_converged::ScalarWrapper{Int}
+    Δt::ScalarWrapper{T}
+end
+
+function AdaptiveTimeStepper(
+    Δt_init::T, t_end::T; 
+    t_start=zero(T), Δt_min=Δt_init, Δt_max=typemax(T), 
+    change_factor=T(1.5), num_converged_to_increase::Int=1) where T
+    
+    num_converged = ScalarWrapper(0)
+    Δt = ScalarWrapper(Δt_init)
+    return AdaptiveTimeStepper(
+        t_start, t_end, Δt_init, Δt_min, Δt_max,
+        change_factor, num_converged_to_increase,
+        num_converged, Δt)
+end
+
+initial_time(ts::AdaptiveTimeStepper) = ts.t_start 
+islaststep(ts::AdaptiveTimeStepper, t, step) = t >= ts.t_end - eps(t)
+function update_time(s::FerriteSolver{<:Any, <:AdaptiveTimeStepper}, t, step, converged)
+    ts=s.timestepper
+    
+    # Initialization
+    if step == 1 
+        ts.Δt[] = ts.Δt_init   
+        ts.num_converged[] = 0
+    end
+
+    if !converged
+        if step == 1
+            msg = "step=1 implies initial step and then \"convergence of the previous step\" must be true"
+            throw(ArgumentError(msg))
+        end
+        if ts.Δt[] ≈ ts.Δt_min
+            msg = "The nonlinear solve failed and the AdaptiveTimeStepper is at its minimum time step"
+            throw(ConvergenceError(msg))
+        end
+        t -= ts.Δt[]
+        ts.Δt[] = max(ts.Δt[]/ts.change_factor, ts.Δt_min)
```

Comment:
If I understand correctly we only shrink the timestep size if we fail to converge, correct?

@lijas  uses a criterion based on the current amount of iterations done in the solver, which I really like, since this opens the possibility to decrease the step size even before convergence fails. Let's say you would specify 5 as optimum newton iteration, this procedure would increase the step size for less than 5 and decrease for more than 6 based on some `2^difference_in_optimum_and_current_iteration` quantity

Reply:
Yes, that is a good point, I'll do that!

Reply:
An issue now is where the settings should belong. Should the nonlinear solver struct have the optimum number of iterations, or should that belong to the adaptive time stepper...
For now, I've made that optional by defaulting to half the number of maxiter, which I assume all nonlinear solvers should have defined, and then one can define this for the specific solver, or even overload it by using a function. I'm not sure what is the best option here (if adding it to the time stepper, one could get issues such as defining maxiter<optimiter that is harder to provide checks for.) Would be happy for suggestions for a nicer solution!